### PR TITLE
[SCM-914] Introduce properly typed last modified date

### PR DESCRIPTION
**Pull request by kwin**
Populate it with SVN, Git and JGit for now

### CODE REVIEW

**Comment by kwin** [comment]
> I would use rather `Instant`. Client code needs to convert to an offset or timezone if required. We should store it neutral.

That would have the disadvantage that the original offset information gets lost. Although for most operations the normalized Instant is enough there might be edge cases where the offset might be interesting. Both Git and Subversion keep the offset as well (and don’t just emit UTC dates)

**Comment by michael-o** [comment]
> > I would use rather `Instant`. Client code needs to convert to an offset or timezone if required. We should store it neutral.
> 
> That would have the disadvantage that the original offset information gets lost. Although for most operations the normalized Instant is enough there might be edge cases where the offset might be interesting. Both Git and Subversion keep the offset as well (and don’t just emit UTC dates)

Subversion does not. It stores Zulu in `svn:date` and then coverts to your time zone. Only Git stores, for a reason I don't understand, the offset to the Unix timestamp.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/info/GitInfoCommand.java [hunk 1]
```diff
@@ -43,31 +50,37 @@ public class GitInfoCommand extends AbstractCommand implements GitCommand {
     protected ScmResult executeCommand(
             ScmProviderRepository repository, ScmFileSet fileSet, CommandParameters parameters) throws ScmException {
 
-        GitInfoConsumer consumer = new GitInfoConsumer(fileSet);
-        CommandLineUtils.StringStreamConsumer stderr = new CommandLineUtils.StringStreamConsumer();
-
-        Commandline cli = createCommandLine(repository, fileSet, parameters);
+        Commandline baseCli = GitCommandLineUtils.getBaseGitCommandLine(fileSet.getBasedir(), "log");
+        baseCli.createArg().setValue("-1"); // only most recent commit matters
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/info/GitInfoCommand.java:54]: Stupid question: Should be exclude merge commits here?

**Review by michael-o — CHANGES_REQUESTED** [review]

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/info/GitInfoCommand.java [hunk 2]
```diff
@@ -43,31 +49,34 @@ public class GitInfoCommand extends AbstractCommand implements GitCommand {
...
-        // different from -1
-        {
-            cli.createArg().setValue("--short=" + revLength);
+    protected InfoItem executeInfoCommand(Commandline cli, CommandParameters parameters, File scmFile)
+            throws ScmException {
+        GitInfoConsumer consumer = new GitInfoConsumer(scmFile.toPath(), getRevisionLength(parameters));
+        CommandLineUtils.StringStreamConsumer stderr = new CommandLineUtils.StringStreamConsumer();
+        int exitCode = GitCommandLineUtils.execute(cli, consumer, stderr);
+        if (exitCode != 0) {
+            throw new ScmException("The git log command failed:" + cli.toString() + " returned " + stderr.getOutput());
```
[the remark this answers is not in the document]
Comment (kwin) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/info/GitInfoCommand.java]: fixed

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/info/GitInfoConsumer.java [hunk 3]
```diff
@@ -18,50 +18,90 @@
...
-        return infoItems;
+    /**
+     * The format argument to use with {@code git log}
+     * @return the format argument to use {@code git log} command
+     * @see <a href="https://git-scm.com/docs/git-log#_pretty_formats">Pretty Formats</a>
+     */
+    public static Arg getFormatArgument() {
+        Commandline.Argument arg = new Commandline.Argument();
+        arg.setValue("--format=format:%H %aI %aE %aN");
+        return arg;
```
[the remark this answers is not in the document]
Comment (kwin) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/info/GitInfoConsumer.java:105]: I enriched the TCK and test with Svnexe, Gitexe and JGit now that parsing works.

File: maven-scm-api/src/main/java/org/apache/maven/scm/command/info/InfoItem.java [hunk 4]
```diff
@@ -117,11 +126,31 @@ public void setLastChangedRevision(String lastChangedRevision) {
...
     public void setLastChangedDate(String lastChangedDate) {
         this.lastChangedDate = lastChangedDate;
     }
+
+    /**
+     *
+     * @return the date when the file indicated via {@link #getPath()} has been changed in the SCM for the last time
+     */
+    OffsetDateTime getLastModifiedDate() {
+        return lastChangedDateTime;
```
[the remark this answers is not in the document]
Comment (kwin) [reply maven-scm-api/src/main/java/org/apache/maven/scm/command/info/InfoItem.java:151]: I renamed to getLastChangedDateTime().

File: maven-scm-api/src/main/java/org/apache/maven/scm/command/info/InfoItem.java [hunk 5]
```diff
@@ -18,7 +18,14 @@
  */
 package org.apache.maven.scm.command.info;
 
+import java.time.OffsetDateTime;
+import java.time.temporal.TemporalAccessor;
+
 /**
+ * Encapsulates meta information about one file (or directory) being managed with an SCM.
```
[the remark this answers is not in the document]
Comment (kwin) [reply maven-scm-api/src/main/java/org/apache/maven/scm/command/info/InfoItem.java]: fixed

**Review by michael-o — CHANGES_REQUESTED** [review]
Can you double check the failing CI tests?

**Comment by kwin** [comment]
It only fails on Windows. I don’t have a VM or computer at hand and the logs from CI are not verbose enough for me. @michael-o Do you have a Windows machine where you could check?

**Comment by michael-o** [comment]
I can reproduce the error locally. Let me check...

**Comment by michael-o** [comment]
Output:
```
-------------------------------------------------------------------------------
Test set: org.apache.maven.scm.provider.git.gitexe.command.info.GitExeInfoCommandTckTest
-------------------------------------------------------------------------------
Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.995 s <<< FAILURE! -- in org.apache.maven.scm.provider.git.gitexe.command.info.GitExeInfoCommandTckTest
org.apache.maven.scm.provider.git.gitexe.command.info.GitExeInfoCommandTckTest.testInfoCommandFromBasedirDifferentFromWorkingCopyDirectory -- Time elapsed: 0.875 s <<< ERROR!
org.apache.maven.scm.ScmException: Exception while executing SCM command.
	at org.apache.maven.scm.command.AbstractCommand.execute(AbstractCommand.java:54)
	at org.apache.maven.scm.provider.git.AbstractGitScmProvider.executeCommand(AbstractGitScmProvider.java:257)
	at org.apache.maven.scm.provider.git.AbstractGitScmProvider.info(AbstractGitScmProvider.java:266)
	at org.apache.maven.scm.tck.command.info.InfoCommandTckTest.testInfoCommandFromBasedirDifferentFromWorkingCopyDirectory(InfoCommandTckTest.java:57)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:316)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:240)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:214)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:155)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
Caused by: org.apache.maven.scm.ScmException: Error while executing command.
	at org.apache.maven.scm.provider.git.gitexe.command.GitCommandLineUtils.execute(GitCommandLineUtils.java:159)
	at org.apache.maven.scm.provider.git.gitexe.command.info.GitInfoCommand.executeInfoCommand(GitInfoCommand.java:78)
	at org.apache.maven.scm.provider.git.gitexe.command.info.GitInfoCommand.executeCommand(GitInfoCommand.java:68)
	at org.apache.maven.scm.command.AbstractCommand.execute(AbstractCommand.java:52)
	... 34 more
Caused by: org.codehaus.plexus.util.cli.CommandLineException: Error while executing process.
	at org.codehaus.plexus.util.cli.Commandline.execute(Commandline.java:598)
	at org.codehaus.plexus.util.cli.CommandLineUtils.executeCommandLineAsCallable(CommandLineUtils.java:128)
	at org.codehaus.plexus.util.cli.CommandLineUtils.executeCommandLine(CommandLineUtils.java:100)
	at org.codehaus.plexus.util.cli.CommandLineUtils.executeCommandLine(CommandLineUtils.java:68)
	at org.apache.maven.scm.provider.git.gitexe.command.GitCommandLineUtils.execute(GitCommandLineUtils.java:157)
	... 37 more
Caused by: java.io.IOException: Cannot run program "/X" (in directory "D:\Entwicklung\Projekte\maven-scm\maven-scm-providers\maven-scm-providers-git\maven-scm-provider-gitexe\target\scm-test\working-copy\src\main"): CreateProcess error=2, Das System kann die angegebene Datei nicht finden
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at java.lang.Runtime.exec(Runtime.java:593)
	at org.codehaus.plexus.util.cli.Commandline.execute(Commandline.java:595)
	... 41 more
Caused by: java.io.IOException: CreateProcess error=2, Das System kann die angegebene Datei nicht finden
	at java.lang.ProcessImpl.create(Native Method)
	at java.lang.ProcessImpl.<init>(ProcessImpl.java:453)
	at java.lang.ProcessImpl.start(ProcessImpl.java:139)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	... 43 more
```

**Comment by michael-o** [comment]
`cmd.exe` is gone from command line:
```
[main] INFO org.apache.maven.scm.provider.git.gitexe.command.GitCommandLineUtils - Executing: /X /C git log -1 --no-merges "--format=format:%H %aI %aE %aN" -- java/Application.java
```

WTF?

**Comment by michael-o** [comment]
Found it: The `clone()` method is broken in `Commandline`/`Shell`.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/info/JGitInfoCommand.java [hunk 6]
```diff
@@ -74,6 +76,8 @@ protected ScmResult executeCommand(
             return new InfoScmResult(infoItems, new ScmResult("JGit.resolve(HEAD)", "", objectId.toString(), true));
         } catch (Exception e) {
             throw new ScmException("JGit resolve failure!", e);
+        } finally {
+            JGitUtils.closeRepo(git);
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/info/JGitInfoCommand.java:80]: This shouldn't make a difference, no?

**Comment by michael-o** [comment]
This PR triggers https://github.com/apache/mina-sshd/issues/282. We first need to upgrade MINA. I am working on this right now.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/info/JGitInfoCommand.java (hunk 6 above)
Comment (kwin) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/info/JGitInfoCommand.java:80]: It does! `git.close()` does not close the underlying repo unfortunately.

**Comment by kwin** [comment]
> This PR triggers [apache/mina-sshd#282](https://github.com/apache/mina-sshd/issues/282). We first need to upgrade MINA. I am working on this right now.

How did this PR trigger this?

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/info/JGitInfoCommand.java (hunk 6 above)
Comment (michael-o) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/info/JGitInfoCommand.java:80]: Insane, this is unlogical.

**Comment by michael-o** [comment]
> > This PR triggers [apache/mina-sshd#282](https://github.com/apache/mina-sshd/issues/282). We first need to upgrade MINA. I am working on this right now.
> 
> How did this PR trigger this?

Not intentionally, I didn't notice it before. As it seems it is also present on master.

**Review by michael-o — APPROVED** [review]
As soon as CI passes this is good to merge.

**Comment by jira-importer** [comment]
Resolve #1134

**Pull request merged**